src/units/drawer.cpp

```diff
@@ -462,6 +469,23 @@ void unit_drawer::redraw_unit(const unit& u) const
 		halo_man.set_location(ac.unit_halo_, halo_x, halo_y);
 	}
 
+	bool has_abil_halo = ac.abil_halo_ && ac.abil_halo_->valid();
+	const std::string& halo_abilities = u.halo_abilities();
+
+	if(!has_abil_halo && !halo_abilities.empty()) {
```

Comment:
There seems to be a missing case here. Assume the unit has two abilities that show different halos, and something triggers the change from one to the other - then `has_abil_halo` will be true, `halo_abilities.empty()` will be false, yet we'll still have the wrong image.

Reply:
missing case fixed

Reply:
Not completely fixed, but possibly only a single-frame glitch.

At the moment, it'll remove the halo instead of replacing it. However, on the next draw it will detect that there's no halo and add the correct one. (Reviewed, but not tested)